<a href="https://colab.research.google.com/github/fabiomiguell/AtividadesMackenzie/blob/main/BIG_A1_LinearRegressionStatsmodels.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<head>
  <meta name="author" content="Rogério de Oliveira">
  <meta institution="author" content="Universidade Presbiteriana Mackenzie">
</head>

<img src="http://meusite.mackenzie.br/rogerio/mackenzie_logo/UPM.2_horizontal_vermelho.jpg" width=300, align="right">
<!-- <h1 align=left><font size = 6, style="color:rgb(200,0,0)"> optional title </font></h1> -->


**Nome**: Fabio de Oliveira Miguel

**Matrícula**: 92274714

**Link Colab**: 

# **Tarefa de Aprofundamento 1** 
## CASO: Estimando o Preço de Imóveis em São Paulo
---
<br>
<br>





A partir deste notebook crie um modelo de regressão múltipla com a biblioteca `statsmodels` para predição de preço de imóveis com base no histórico de preços e características de vários imóveis de São Paulo. 

Em seguida, com base nos seus resultados, responda o **questionário do Moodle**. 

# Basic imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import statsmodels.formula.api as sm

# Exploração dos Dados

Entenda os dados disponíveis e em seguida faça as transformações necessárias nos dados. Quantidades dos dados, se existem valores nulos e valores máximos e mínimos dos atributos são algumas explorações comuns. 

(Moodle) **Q1**. Quantos bairros existem no histórico e qual o bairro mais presente nos dados?



In [2]:
df = pd.read_excel('http://meusite.mackenzie.br/rogerio/data_load/regressao_preco_imoveis.xlsx')
df.head()

bairro  areaM2  suites  dormitorios  banheiros  vagas    preco
0  vila-nova-conceicao      32       1            1          1      1   490000
1  vila-nova-conceicao     157       2            2          2      2  3180000
2  vila-nova-conceicao     205       2            3          3      3  1900000
3  vila-nova-conceicao     193       3            3          3      3  3565000
4  vila-nova-conceicao     116       1            3          2      2  1605000

In [6]:
df.columns = [ x.replace('.','') for x in df.columns ]
df.head()

bairro  areaM2  suites  dormitorios  banheiros  vagas    preco
0  vila-nova-conceicao      32       1            1          1      1   490000
1  vila-nova-conceicao     157       2            2          2      2  3180000
2  vila-nova-conceicao     205       2            3          3      3  1900000
3  vila-nova-conceicao     193       3            3          3      3  3565000
4  vila-nova-conceicao     116       1            3          2      2  1605000

In [7]:
df.describe(include='all')

bairro       areaM2       suites  dormitorios    banheiros  \
count           3741  3741.000000  3741.000000  3741.000000  3741.000000   
unique             5          NaN          NaN          NaN          NaN   
top     vila-mariana          NaN          NaN          NaN          NaN   
freq            1714          NaN          NaN          NaN          NaN   
mean             NaN   140.343758     1.700882     2.946271     3.095429   
std              NaN    60.051791     0.994945     0.759292     1.372859   
min              NaN    20.000000     1.000000     1.000000     1.000000   
25%              NaN    94.000000     1.000000     3.000000     2.000000   
50%              NaN   130.000000     1.000000     3.000000     3.000000   
75%              NaN   180.000000     2.000000     3.000000     4.000000   
max              NaN   299.000000     6.000000     6.000000     9.000000   

              vagas         preco  
count   3741.000000  3.741000e+03  
unique          NaN           NaN  
top             NaN           NaN  
freq            NaN           NaN  
mean       2.225341  1.553746e+06  
std        1.000940  1.073607e+06  
min        1.000000  2.800000e+05  
25%        1.000000  8.500000e+05  
50%        2.000000  1.200000e+06  
75%        3.000000  1.880000e+06  
max        5.000000  8.000000e+06

**Resposta Q1**: a base possui 3741 bairros e Vila Mariana é o mais presente com 1714 registros


# Preparação dos Dados

Use [statsmodels.formula.api](https://www.statsmodels.org/dev/generated/statsmodels.formula.api.ols.html)

```
from statsmodels.formula.api import ols   
```

Se você empregar esta forma do `statsmodels` ele fará de modo implítico o hot encode dos dados e não há necessidade de nenhuma transformação dos dados (não faremos a normalização aqui. Embora sempre recomendável na maior parte dos casos, aqui o resultado de predição será o mesmo com o sem a normalização).

Atenção, se você empregar 
```
from statsmodels.api import OLS 
```
os resultados serão os mesmos, mas alguma preparação nos dados será necessária e a construção do modelo será um pouco diferente.

In [9]:
from statsmodels.formula.api import ols   

# nada a fazer se estiver empregando a api acima

# Estimando o modelo com o `statsmodels`

Estime o modelo.

(Moodle) **Q2**. Qual o coeficiente de determinação ajustado obtido?

Dica: empregue `result.rsquared_adj` no lugar de `result.summary()` para obter resultados com maior precisão.

(Moodle) **Q3**. Qual coeficiente estimado é menos significativo?

In [15]:
model_1 = sm.ols("preco ~ areaM2 + suites + dormitorios + banheiros + vagas + bairro", data=df)
result_1 = model_1.fit()
print(result_1.rsquared_adj)

0.7948154407290992


**Resposta Q2**: O coeficiente de determinação ajustado é 0.79481

In [17]:
model_1 = sm.ols("preco ~ areaM2 + suites + dormitorios + banheiros + vagas + bairro", data=df)
result_1 = model_1.fit()
print(result_1.rsquared_adj)
print(result_1.summary())

0.7948154407290992
                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     1611.
Date:                Sat, 27 May 2023   Prob (F-statistic):               0.00
Time:                        08:59:55   Log-Likelihood:                -54290.
No. Observations:                3741   AIC:                         1.086e+05
Df Residuals:                    3731   BIC:                         1.087e+05
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------

**Resposta Q3**: O coeficiente estimado menos significativo é banheiro, que apresentou resultado P>|t| mais próximo de 1

# Re-estimando o modelo com o `statsmodels`

Estime novamente o modelo excluindo o atributo menos significativo (um único atributo). 

(Moodle) **Q4**. Qual o novo coeficiente de determinação ajustado obtido?


In [14]:
model_2 = sm.ols(formula="preco ~ areaM2 + suites + dormitorios + vagas + bairro - 1", data=df)
result_2 = model_2.fit()
print(result_2.rsquared_adj)

0.7948526838722687


**Resposta Q4**: O novo coeficiente de determinação ajustado é 0.79485, ou seja, a retirada do atributo melhorou o modelo.

In [19]:
print(result_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     1812.
Date:                Sat, 27 May 2023   Prob (F-statistic):               0.00
Time:                        09:09:48   Log-Likelihood:                -54290.
No. Observations:                3741   AIC:                         1.086e+05
Df Residuals:                    3732   BIC:                         1.087e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
bairro[morumbi]   

# Predição

Faça a predição do preço, com o melhor dos modelos obtidos acima, de um imóvel na Vila Mariana, com 134 m2, 1 suite, 4 dorm, 3 banheiros e 3 vagas.



(Moodle) **Q5**. Qual o preço estimado, para o melhor dos modelos obtidos, de um imóvel na Vila Mariana, com 134 m2, 1 suite, 4 dorm, 3 banheiros e 3 vagas?



In [13]:
x = pd.DataFrame({'bairro':['vila-mariana'], 'areaM2':[134], 'suites':[1], 'dormitorios':[4], 'banheiros':[3], 'vagas':[3]})
result_2.predict(x)

0    1.350585e+06
dtype: float64

**Resposta Q5**: O preço estimado para as características solicitadas é R$ 1.350.585,00